tests/queries/0_stateless/02346_text_index_preprocessor.sql

```diff
@@ -0,0 +1,165 @@
+SELECT count() FROM tab WHERE hasToken(str, 'FOO');
+
+SELECT count() FROM tab WHERE hasToken(str, 'BAR');
+SELECT count() FROM tab WHERE hasToken(str, 'Baz');
+
+SELECT count() FROM tab WHERE hasToken(str, 'bar');
+SELECT count() FROM tab WHERE hasToken(str, 'baz');
+
+DROP TABLE tab;
+
+SELECT 'Test preprocessor declaration using column more than once.';
+CREATE TABLE tab
+(
+    key UInt64,
+    str String,
+    INDEX idx(str) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = concat(str, str))
+)
+ENGINE = MergeTree
+ORDER BY tuple();
+
+
+INSERT INTO tab VALUES (1, 'foo'), (2, 'BAR'), (3, 'Baz');
+
+SELECT count() FROM tab WHERE hasToken(str, 'foo');
+SELECT count() FROM tab WHERE hasToken(str, 'FOO');
+
+SELECT count() FROM tab WHERE hasToken(str, 'BAR');
+SELECT count() FROM tab WHERE hasToken(str, 'Baz');
+
+SELECT count() FROM tab WHERE hasToken(str, 'bar');
+SELECT count() FROM tab WHERE hasToken(str, 'baz');
+
+DROP TABLE tab;
+
+SELECT 'Negative tests on preprocessor construction validations.';
+
+-- The preprocessor argument must reference the index column
+CREATE TABLE tab
+(
+    key UInt64,
+    str String,
+    other_str String,
+    INDEX idx(str) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = lower(other_str))
+)
+ENGINE = MergeTree ORDER BY tuple();  -- { serverError BAD_ARGUMENTS }
+
+-- The preprocessor argument must not reference non-indexed columns
+CREATE TABLE tab
+(
+    key UInt64,
+    str String,
+    other_str String,
+    INDEX idx(str) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = concat(str, other_str))
+)
+ENGINE = MergeTree ORDER BY tuple();   -- { serverError BAD_ARGUMENTS }
+
+-- The preprocessor argument must contain the index expression
+CREATE TABLE tab
+(
+    key UInt64,
+    str String,
+    INDEX idx(upper(str)) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = lower(str))
+)
+ENGINE = MergeTree ORDER BY tuple();   -- { serverError BAD_ARGUMENTS }
+
+CREATE TABLE tab
+(
+    key UInt64,
+    str String,
+    INDEX idx(lower(str)) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = lower(str))
+)
+ENGINE = MergeTree ORDER BY tuple();   -- { serverError BAD_ARGUMENTS }
+
+CREATE TABLE tab
+(
+    key UInt64,
+    str String,
+    INDEX idx(upper(str)) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = lower(upper(str)))
+)
+ENGINE = MergeTree ORDER BY tuple();   -- { serverError BAD_ARGUMENTS }
+
+-- The preprocessor argument must reference actual functions
+CREATE TABLE tab
+(
+    key UInt64,
+    str String,
+    INDEX idx(str) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = nonExistingFunction)
+)
+ENGINE = MergeTree ORDER BY key;   -- { serverError INCORRECT_QUERY }
+
+
+CREATE TABLE tab
+(
+    key UInt64,
+    str String,
+    INDEX idx(str) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = nonExistingFunction(str))
+)
+ENGINE = MergeTree ORDER BY tuple();   -- { serverError UNKNOWN_FUNCTION }
```

Comment:
Minor: If the error is from "our" code, would it be possible to throw `BAD_ARGUMENTS`? Same in l. 110. and l. 128.

Reply:
Added UDF test and it works.